README.md
# Nonogram

[![demo](http://i.imgur.com/XRs3jk7.gif)](https://handsomeone.github.io/Nonogram)

## Usage

[**Check the live demo and find out how to build your own nonogram application.**](https://handsomeone.github.io/Nonogram)

You just need to attach

```html
<script src="https://cdn.rawgit.com/HandsomeOne/Nonogram/gh-pages/nonogram.min.js"></script>
```

to `<head>`. A `<canvas>` element is required for each nonogram instance.

## API

### `class nonogram.Solver`

#### `#constructor(row, column, canvas[, config])`

Creates a nonogram solver.

- `row`: a two-dimensional array, consisting of the hints of each row as an array.
- `column`: a two-dimensional array, consisting of the hints of each column as an array.
- *optional* `canvas`: a canvas element, or `id` of the canvas to print the nonogram on. If not given, a new canvas element will be created and assigned to `this.canvas` so you can put it to the document later.
- *optional* `config`: an object, see [§ Configuration Items](#configuration-items).

#### `#solve()`

Solves and prints the nonogram by given hints.

For example, if there is `<canvas id="canvas1"></canvas>`, then you can use
```javascript
var s = new nonogram.Solver(
  [
    [1, 1],
    [1, 1],
    [1, 1],
    [4]
  ],
  [
    [4],
    [1],
    [1],
    [4]
  ],
  'canvas1',
  { width: 500, delay: 100 }
)
s.solve()
```
then the output will be like this:
```
██    ██ 1 1
██    ██ 1 1
██    ██ 1 1
████████ 4
4 1 1 4
```

### `class nonogram.Editor`

#### `#constructor(m, n, canvas[, config])`

Creates a nonogram editor.

- `m`: number of rows, or the length of each column.
- `n`: number of columns, or the length of each row.
- *optional* `canvas`: same as that of `nonogram.Solver`.
- *optional* `config`: an object, see [§ Configuration Items](#configuration-items).

#### `#refresh()`

Randomly generates the grid.

For example, if you run
```javascript
new nonogram.Editor(4, 6, 'canvas2', {threshold: 0.9})
```
then the output is likely to be
```
████████████ 6
████  ██████ 2 3
  ██████████ 5
████████████ 6
2 4 1 4 4 4
1   2
```

### `class nonogram.Game`

#### `#constructor(row, column, canvas[, config])`

Creates a nonogram game. The parameters have the same definitions as those of `nonogram.Solver`'s.

## Configuration Items

- `theme`: an plain object, controls the appearance.
  - `width` (px): a number to set the canvas' width. If not given, the canvas' current `clientWidth` (**not** the value of its `width` property) will be used.
  - `filledColor`: filled cells' color.
  - `unsetColor`: unset cells' color.
  - `correctColor`: numbers' color of correct rows or columns.
  - `wrongColor`: numbers' color of wrong rows or columns.
  - `meshColor`: meshes' color.
  - `isMeshed`: `true` or `false`, coltrols whether to print the meshes or not.
  - `isBoldMeshOnly`: default is `false`.
  - `isMeshOnTop`: default is `false`.
  - `boldMeshGap`: default is `5`. Controls how many cells are there between two adjacent bold meshes. If you don't want any bold meshes, simply set it to `0`.

### `nonogram.Solver`

- `delay` (ms): default is `50`. Controls the delay between steps of the solving process.

- `onSuccess(time)`: fired when the nonogram has been solved, `time` is how many milliseconds cost.

- `onError(err)`: when some contradiction has been found, `err` tells the bad hints' location (index starts at 1).

### `nonogram.Editor`

- `grid`: a two-dimensional array, consisting of `1`s and `0`s, will be assigned to the nonogram's grid. For example, you can use
```javascript
[[1, 0, 0, 1],
[1, 0, 0, 1],
[1, 0, 0, 1],
[1, 1, 1, 1]]
```
to create
```
██    ██ 1 1
██    ██ 1 1
██    ██ 1 1
████████ 4
4 1 1 4
```

- `threshold`: if `grid` is not given, then the nonogram's grid will be randomly generated. Each cell of the grid has a chance of threshold*100% to be filled. Default is `0.5`.

- `onHintChange(row, column)`: fired when the nonogram's hints have any change. To automatically create a new solver on hint change, you can use
```javascript
new nonogram.Editor(4, 4, 'canvas1', {
  onHintChange: function (row, column) {
    new nonogram.Solver(row, column, 'canvas2').solve()
  })
})
```

### `nonogram.Game`

- `onSuccess()`: fired when the player has successfully solved the nonogram.

- `onAnimationEnd()`: fired when when the success animation has been finished.


CHANGELOG.md
# Change log

### 0.1.0

- start using babel and webpack

### 0.2.0

- constructors are using namespace `nonogram`
- use truthy and falsy values to setting `Editor`s grid

### 0.3.0

- removed custom event, using callbacks instead

### 0.4.0

- canvas in constructors goes optional
- some configurable items wrapped in key `theme`

### 0.5.0

- now using TypeScript

### 0.6.0

- now using `WebWorker`

### 0.7.0

- enhance `worker.ts`
- remove `demoMode`, `!!delay` now has the same meaning

### 0.8.0

- use rollup instead of webpack


src/Solver.ts
import Nonogram from './Nonogram'
import $ from './colors'
import Status from './status'

import SolverWorker from 'worker!./worker.ts'

export default class Solver extends Nonogram {
  worker: Worker = new SolverWorker()

  delay: number
  handleSuccess: (time: number) => void
  handleError: (e: Error) => void
  isBusy: boolean
  isError: boolean
  scanner?: {
    direction: Direction
    i: number
  }
  startTime: number

  constructor(
    row: number[][],
    column: number[][],
    canvas: string | HTMLCanvasElement,
    {
      theme = {},
      delay = 50,
      onSuccess = () => { },
      onError = () => { },
    } = {},
  ) {
    super()
    this.theme.filledColor = $.green
    this.theme.correctColor = $.green
    this.theme.wrongColor = $.yellow
    Object.assign(this.theme, theme)

    this.delay = delay
    this.handleSuccess = onSuccess
    this.handleError = onError

    this.hints = {
      row: row.slice(),
      column: column.slice(),
    }
    this.removeNonPositiveHints()
    this.m = this.hints.row.length
    this.n = this.hints.column.length
    this.grid = new Array(this.m)
    for (let i = 0; i < this.m; i += 1) {
      this.grid[i] = new Array(this.n).fill(Status.UNSET)
    }

    this.initCanvas(canvas)
    this.print()
  }

  initListeners() {
    this.listeners = [
      ['click', this.click],
    ]
  }
  click = (e: MouseEvent) => {
    if (this.isBusy) return

    const rect = this.canvas.getBoundingClientRect()
    const x = e.clientX - rect.left
    const y = e.clientY - rect.top
    const d = rect.width * 2 / 3 / (this.n + 1)
    const location = this.getLocation(x, y)
    if (location === 'grid') {
      if (this.isError) return

      const i = Math.floor(y / d - 0.5)
      const j = Math.floor(x / d - 0.5)
      if (this.grid[i][j] === Status.UNSET) {
        this.grid[i][j] = Status.FILLED
        this.hints.row[i].unchanged = false
        this.hints.column[j].unchanged = false
        this.solve()
      }
    } else if (location === 'controller') {
      this.refresh()
    }
  }
  private refresh() {
    this.grid = new Array(this.m)
    for (let i = 0; i < this.m; i += 1) {
      this.grid[i] = new Array(this.n).fill(Status.UNSET)
    }
    this.hints.row.forEach((r) => {
      r.isCorrect = false
      r.unchanged = false
    })
    this.hints.column.forEach((c) => {
      c.isCorrect = false
      c.unchanged = false
    })
    this.solve()
  }
  solve() {
    if (this.isBusy) return

    this.print()
    this.isBusy = true
    this.startTime = Date.now()
    this.worker.onmessage = ({ data }: {data: SolverMessage}) => {
      if (this.canvas.nonogram !== this) {
        this.worker.terminate()
        return
      }

      this.scanner = data.scanner
      this.grid = data.grid
      this.hints = data.hints
      if (data.type !== 'update') {
        this.isBusy = false
        if (data.type === 'error') {
          this.isError = true
          const { direction, i } = <Scanner>this.scanner
          this.handleError(new Error(
            `Bad hints at ${direction} ${i + 1}`
          ))
        } else if (data.type === 'finish') {
          this.isError = false
          this.handleSuccess(Date.now() - this.startTime)
        }
      }
      this.print()
    }
    this.worker.postMessage({
      delay: this.delay,
      grid: this.grid,
      hints: this.hints,
    })
  }

  print() {
    this.printGrid()
    this.printHints()
    this.printScanner()
    this.printController()
  }
  printController() {
    const { ctx } = this
    const { width: w, height: h } = this.canvas
    const controllerSize = Math.min(w, h) / 4
    const filledColor = this.theme.filledColor

    function getCycle() {
      const cycle = document.createElement('canvas')
      const borderWidth = controllerSize / 10
      cycle.width = controllerSize
      cycle.height = controllerSize

      const c = cycle.getContext('2d') || new CanvasRenderingContext2D()
      c.translate(controllerSize / 2, controllerSize / 2)
      c.rotate(Math.PI)
      c.arc(0, 0, controllerSize / 2 - borderWidth / 2, Math.PI / 2, Math.PI / 3.9)
      c.lineWidth = borderWidth
      c.strokeStyle = filledColor
      c.stroke()
      c.beginPath()
      c.moveTo((controllerSize / 2 + borderWidth) * Math.SQRT1_2,
        (controllerSize / 2 + borderWidth) * Math.SQRT1_2)
      c.lineTo((controllerSize / 2 - borderWidth * 2) * Math.SQRT1_2,
        (controllerSize / 2 - borderWidth * 2) * Math.SQRT1_2)
      c.lineTo((controllerSize / 2 - borderWidth * 2) * Math.SQRT1_2,
        (controllerSize / 2 + borderWidth) * Math.SQRT1_2)
      c.closePath()
      c.fillStyle = filledColor
      c.fill()

      return cycle
    }

    ctx.clearRect(w * 2 / 3 - 1, h * 2 / 3 - 1, w / 3 + 1, h / 3 + 1)
    if (this.isBusy) return

    ctx.save()
    ctx.translate(w * 0.7, h * 0.7)
    ctx.drawImage(getCycle(), 0, 0)
    ctx.restore()
  }
  printScanner() {
    if (!this.scanner) return

    const { ctx } = this
    const { width: w, height: h } = this.canvas
    const d = w * 2 / 3 / (this.n + 1)

    ctx.save()
    ctx.translate(d / 2, d / 2)
    ctx.fillStyle = this.isError ? this.theme.wrongColor : this.theme.correctColor
    ctx.globalAlpha = 0.5
    if (this.scanner.direction === 'row') {
      ctx.fillRect(0, d * this.scanner.i, w, d)
    } else if (this.scanner.direction === 'column') {
      ctx.fillRect(d * this.scanner.i, 0, d, h)
    }
    ctx.restore()
  }
}
